NE-1325: External DNS Operator support for Shared VPCs (#1436)
openshift/enhancements: enhancements/dns/externaldns-operator-aws-assume-role.md

    // +optional
    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
}

type ExternalDNSAWSAssumeRoleOptions struct {
    // arn is an AWS role ARN that the ExternalDNS
    // operator will assume when making DNS updates.
    //
    // +kubebuilder:validation:Required
    // +required
    ARN string `json:"arn,omitempty"`
}

```

### Implementation Details/Notes/Constraints

#### The Required Credentials Issue

Extending the v1beta1 `ExternalDNS` API object revealed a pre-existing issue with the `Credentials` API field in
`ExternalDNSAWSProviderOptions`. The problem is that `Credentials` is designated as required, even though in the context
of OpenShift, it is not actually necessary. This is because the ExternalDNS Operator generates a `CredentialRequest`
automatically. This wasn't yet exposed as an issue because there were no other fields in `spec.provider.aws`, and
`spec.provider.aws` is optional. The underlying [logic](https://github.com/openshift/external-dns-operator/blob/11e3b72b75d696b7419bce8360183373fc0725e1/api/v1beta1/externaldns_webhook.go#L132)
of the webhook is accurate; the issue lies with this required field in the `ExternalDNS` CRD.

> The appropriate fix for this issue is to:

Author: Yea that's a fair point. Done.

    // any DNS updates.
    //
    // +kubebuilder:validation:Optional
    // +optional
    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
}

type ExternalDNSAWSAssumeRoleOptions struct {
    // arn is an AWS role ARN that the ExternalDNS
    // operator will assume when making DNS updates.
    //
    // +kubebuilder:validation:Required
    // +required
    ARN string `json:"arn,omitempty"`
}

```

### Implementation Details/Notes/Constraints

#### The Required Credentials Issue

Extending the v1beta1 `ExternalDNS` API object revealed a pre-existing issue with the `Credentials` API field in
`ExternalDNSAWSProviderOptions`. The problem is that `Credentials` is designated as required to satisfy vanilla
Kubernetes use cases. On OpenShift, the ExternalDNS Operator generates a `CredentialRequest` automatically, eliminating
the need to manually specify `Credentials`. This wasn't yet exposed as an issue because there were no other fields in
`spec.provider.aws`, and `spec.provider.aws` is optional. The underlying [logic](https://github.com/openshift/external-dns-operator/blob/11e3b72b75d696b7419bce8360183373fc0725e1/api/v1beta1/externaldns_webhook.go#L132)

> of the webhook is accurate; the issue lies with this required field in the `ExternalDNS` CRD.

Reviewer: What was the issue?  They are required now but weren't before?

However, there exists a simple workaround: OpenShift users can specify an empty `""` `Credentials` field as follows:
```yaml
apiVersion: externaldns.olm.openshift.io/v1beta1
kind: ExternalDNS
metadata:
  name: sample-aws
spec:
  provider:
    type: AWS
    aws:
      assumeRole:
        arn: "arn:aws:xxx"
      credentials:
        name: ""
```
This satisfies the `ExternalDNS` CRD `+required` field while the ExternalDNS Operator logic [ignores](https://github.com/openshift/external-dns-operator/blob/bac092e98fe5f9065b75bd0fa21d5aeff9d00853/pkg/operator/controller/credentials-secret/controller.go#L259-L267)
the empty string and uses the `CredentialsRequest`. As a result, we plan to introduce this `AssumeRole` field in the
v1beta1 release accompanied by documentation that specifies this requirement to specify an empty `Credentials` field in
OpenShift.

> In the next version of ExternalDNS Operator's API, we will address the requirement for the `Credentials` field as

Reviewer: Unless users might be filling this with a real value now, I think it might be a better idea to start with the issue fixed and transparent to the user. Maybe write the code that fills credentials automatically when needed (or default to "" now and change the default later if needed?).

Author: I've been looking into what we could do to fill it with a real or default value.

I have three ideas I've explored:
1. Add `// +kubebuilder:default:={"name":""}` to the `Credentials` field v1beta1 CRD so it auto fills the default value
    - This works, but might break the upgrade path as laid out by the API team. I don't know if we can implement a CRD default, then convert to a pointer according to [backward-compatibility-gotchas correctly](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas). I'm checking with the API team.
2. Add code in ExternalDNS reconcile loop to add the default values when appropraite
    - This doesn't work because the reconcile loop happens after all validation (webooks, CRD validation  The CR can't be created then updated because it needs to pass validation.
3. Add a mutating webhook to default credentials to `""` when appropriate.
    - This is essentially the same outcome as # 1, but much more painful to implement, we don't have a mutating webhook and would need to build one just for this.

I'm following up with Andrey and the API team based on this suggestion. It's a valid concern, but a complicated one that'd we'd need concurrence from the API team to ensure we don't make our lives unnecessarily harder in the future. Hence, just telling the user to provide an empty credentials seemed the simplest answer,  but I agree that it's not the best UX.

Reference sequence for API validation: 
![image](https://github.com/openshift/enhancements/assets/4316443/bd2a831d-4bdf-4ec1-ba2a-3b054fcdfa35)

Author: **Update**: I finally got a [response](https://redhat-internal.slack.com/archives/CE4L0F143/p1695164515386059?thread_ts=1690552114.405509&cid=CE4L0F143) from the API team. They OK-ed the idea of adding a CRD default of `""` for the `Credentials` field, so this is a major improvement for user experience, making it seamless to the user.

Good idea to pushing for this, it was worth an ask from the API team, they seemed pleasantly shocked by the idea. The implementation update for defaulting is in this PR: https://github.com/openshift/external-dns-operator/pull/199. 

Enhancement Compare Link: https://github.com/openshift/enhancements/compare/b6c090e21f8a653c44c4f5c35faa1f5cf9af61e4..0a508b1baf394bf8fdb45efe2b030f0e4399fbd1

outlined in the alternative [Bumping the API](#bumping-the-api), thus resolving this particular issue.

### Risks and Mitigations

* Introducing this API change in v1beta1 exposes the issue outlined in [The Required Credentials Issue](#the-required-credentials-issue)
  and has the potential to cause confusion. We plan to mitigate this by documenting the limitation. 

### Drawbacks

> No obvious drawbacks as this enhancement is merely exposing an existing ExternalDNS functionality via an API.

Reviewer: What if this is misconfigured or hacked for an unwante privilege escalation?  That would be a big drawback.  General security implications aren't discussed here and probably should be - at least to say that it isn't a concern for reasons detailed in the discussion.

## Design Details

### Open Questions

N/A

### Test Plan

We will add E2E tests for the ExternalDNS Operator utilizing two AWS CI accounts to create DNS records in one (Account
A) and install the cluster in the other (Account B). This E2E test will leverage existing steps in the step registry
as created by [AWS Shared VPC with Cross-account DNS Zones](/enhancements/installer/aws-cross-account-dns-zone.md) such
as [ipi-aws-pre-shared-vpc-phz](https://steps.ci.openshift.org/chain/ipi-aws-pre-shared-vpc-phz) and [ipi-aws-post-shared-vpc-phz](https://steps.ci.openshift.org/chain/ipi-aws-post-shared-vpc-phz).
These steps set up a shared VPC cluster with access to a shared DNS private hosted zone. However, we may choose to use
a public hosted zone instead, as there is no functional distinction between the two, and it simplifies the E2E testing
process.

### Graduation Criteria

This update is an extension of the existing v1beta1 ExternalDNS Operator API and does not impact the graduation
criteria for the v1beta1 API.

The ExternalDNS Operator updates to support Shared VPC are out of payload and are therefore not aligned with

> OpenShift releases. However, we are targeting to release this enhancement of ExternalDNS as alongside of OCP 4.14.

Reviewer: This has changed, right?  It would have to be released in 4.15 and backported to 4.14.z.

would be as follows:
1. Update the CRD schema to make the existing `Credentials` field optional.
2. Create a new version of the API that makes the optional field a pointer (assuming `Credentials` unset and `AssumeRole`
   set is a valid combination).
3. Update new API to not allow zero value of `Credentials`.
4. Make the conversion webhook convert new/nil to old/zero-value and old/zero-value to new/nil, and new/zero-value to
   old/zero-value.

However, the concern with bumping the API solely for this update is that the process of bumping can be quite
time-consuming, and it doesn't align the timeline of this enhancement. Additionally, it would be more efficient to
bundle this fix with other API updates, such as the need to rebase for incorporating an upstream fix related to the
Infoblox view parameter (https://github.com/kubernetes-sigs/external-dns/pull/3301) at a later point.

### Assume Role Strategy

Assuming a different AWS role could be implemented without ExternalDNS's built in `--aws-assume-role` as demonstrated in
https://github.com/openshift/external-dns-operator/pull/191 by using kiam, IRSA, or kube2iam. These are alternate

> methods for assuming a different AWS role, which include utilizing third-party software rather than interacting with
> ExternalDNS directly.

Reviewer: ```suggestion
methods for assuming a different AWS role, which require utilizing third-party software rather than interacting with ExternalDNS directly.  Therefore this method was not chosen.
```

  pre-existing hosted zone.

### Goals

* Allow ExternalDNS Operator users to assume a different role in order to manage DNS records in a pre-existing hosted
  zone that doesn't belong to them.

### Non-Goals

* Support [IRSA](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md#iam-roles-for-service-accounts),
  [kiam](https://github.com/uswitch/kiam), or [kube2Iiam](https://github.com/jtblin/kube2iam) as mechanisms for assuming
  a different role.
* Support specifying of [ExternalID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).
* Provide support for the automatic creation of DNS hosted zones by utilizing the assume role ARN.

## Proposal

Conveniently, ExternalDNS already supports the [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/docs/faq.md#can-external-dns-manageaddremove-records-in-a-hosted-zone-which-is-setup-in-different-aws-account)
argument, which uses the specified AWS role ARN when creating new DNS records. Therefore, to support cross-account DNS hosted

> zones with the ExternalDNS Operator, update the [API](#API-External-DNS-Operator) to expose the `--aws-assume-role`

Reviewer: The API link doesn't seem right.

Author: Good catch, fixed in next push.

information regarding this update, please refer to [The Required Credentials Issue](#the-required-credentials-issue).

```go
type ExternalDNSAWSProviderOptions struct {
    // Credentials is a reference to a secret containing
    // the following keys (with corresponding values):
    //
    // * aws_access_key_id
    // * aws_secret_access_key
    //
    // See
    // https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md
    // for more information.
    //
    // +kubebuilder:validation:Required
    // +kubebuilder:default:={"name":""}
    // +required
    Credentials SecretReference `json:"credentials"`

    // assumeRole is a reference to the IAM role that
    // ExternalDNS will be assuming in order to perform
    // any DNS updates.
    //
    // +kubebuilder:validation:Optional
    // +optional
    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
}

type ExternalDNSAWSAssumeRoleOptions struct {
    // arn is an IAM role ARN that the ExternalDNS
    // operator will assume when making DNS updates.
    //
    // +kubebuilder:validation:Required
    // +required

>     ARN string `json:"arn,omitempty"`

Reviewer: It occurs to me that we may need to plan for multiple ARNs.  Is there any reason to leave this as a single-value field?

Author: Fair question. An ARN represents another IAM Role in another AWS account. If there were multiple ARNs, it would imply that External DNS should assume multiple roles. I can only think of 1 reason to assume multiple roles: to create the same DNS record in different accounts.

The correct answer to satisfy the motivation of creating the same DNS record in multiple AWS accounts with multiple RoleARNs would be: create multiple `ExternalDNS` objects so you have multiple instances of External DNS running to service multiple AWS accounts. I don't know of a use case this would benefit, but my point being, we can still support multiple AWS Accounts (i.e. multiple RoleARN) through multiple instances.

Furthermore, the design of External DNS's argument [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/pkg/apis/externaldns/types.go#L440) doesn't support multiple ARNs. And on top of that, OpenShift is designed to work with [one RoleARN](https://github.com/openshift/enhancements/blob/5cd4edc75ad6144a0b751cb9700d421b5cce217a/enhancements/installer/aws-cross-account-dns-zone.md?plain=1#L138) provided at install time.

I think it's safe to say we don't need to plan for multiple ARNs. Let me know you have information that would make this invalid.

the need to manually specify `Credentials`. This wasn't yet exposed as an issue because there were no other fields in
`spec.provider.aws`, and `spec.provider.aws` is optional. The underlying [logic](https://github.com/openshift/external-dns-operator/blob/11e3b72b75d696b7419bce8360183373fc0725e1/api/v1beta1/externaldns_webhook.go#L132)
of the webhook is accurate; the issue lies with this required field in the `ExternalDNS` CRD.

However, there exists a simple workaround: we can apply a default value of an empty string `""` to the
`spec.aws.provider.credentials.name` field using the `+kubebuilder:default:={"name":""}` specifier on the `Credentials`
field. When `spec.aws.provider` is created, but `spec.aws.provider.credentials` is not specified, it will default
`spec.aws.provider.credentials.name` to `""`.

This satisfies the `+required` field for `Credentials` while the ExternalDNS Operator logic [ignores](https://github.com/openshift/external-dns-operator/blob/bac092e98fe5f9065b75bd0fa21d5aeff9d00853/pkg/operator/controller/credentials-secret/controller.go#L259-L267)

> the empty string and uses the `CredentialsRequest`. Conversely, vanilla Kubernetes users (who do not have the option to
> use a `CredentialsRequest`) will encounter a validation webhook failure when the `Credentials` default of `""` is used
> due to the [validateProviderCredentials](https://github.com/openshift/external-dns-operator/blob/4cb22f19c5e6edcc6a37b73e17f25d996dbef9cd/api/v1beta1/externaldns_webhook.go#L132-L160). This is desirable because non-OpenShift users must specify valid `Credentials`.
> This failure is desirable because vanilla Kubernetes users are required to specify valid `Credentials`.

Reviewer: Is this more accurate?
```suggestion
the empty string and uses the `CredentialsRequest`.  This is desirable because non-OpenShift users must specify valid `Credentials`.  

Conversely, vanilla Kubernetes users (who do not have the option to
use a `CredentialsRequest`) will encounter a validation webhook failure when the `Credentials` default of `""` is used
due to the [validateProviderCredentials](https://github.com/openshift/external-dns-operator/blob/4cb22f19c5e6edcc6a37b73e17f25d996dbef9cd/api/v1beta1/externaldns_webhook.go#L132-L160). 
This failure is desirable because vanilla Kubernetes users are required to specify valid `Credentials`.
```

Author: I'm not sure. The two sentences:
```
This satisfies the +required field for Credentials while the ExternalDNS Operator logic ignores the empty string and uses the CredentialsRequest. This is desirable because non-OpenShift users must specify valid Credentials. 
``` 
The second sentence seems confusing to me. The first was talking in regards to satisfying the OpenShift use-case, but the second talks about the non-OpenShift use case.

I'll make a tweak so this is clearer and less ambiguous. One paragraph about OpenShift users, and the other about non-OpenShift users so it's easier to read.

In a future version of ExternalDNS Operator's API, we will resolve the requirement for the `Credentials` field through a
more appropriate solution as outlined in the alternative [Bumping the API](#bumping-the-api).

### Risks and Mitigations

N/A

### Drawbacks

A minor drawback is the added complexity in supporting cross-account DNS hosted zones. This complexity involves handling
AWS resource sharing, managing IAM policies and roles, and understanding shared VPC architecture.

In terms of security concerns, the IAM role is not considered sensitive because AWS incorporates security measures to
limit the usage of a role ARN to a particular AWS account. Exploitation of the `AssumeRole` field isn't a concern due to

> these AWS security constraints.

Reviewer: I'm still a bit unconvinced that adding this functionality doesn't introduce new security concerns, but I can let it go.

Author: I suppose it's worded too strongly, I'll remove.

would be as follows:
1. Update the CRD schema to make the existing `Credentials` field optional.
2. Create a new version of the API that makes the optional field a pointer (assuming `Credentials` unset and `AssumeRole`
   set is a valid combination).
3. Update new API to not allow zero value of `Credentials`.
4. Make the conversion webhook convert new/nil to old/zero-value and old/zero-value to new/nil, and new/zero-value to
   old/zero-value.

However, the concern with bumping the API solely for this update is that the process of bumping can be quite
time-consuming, and it doesn't align the timeline of this enhancement. Additionally, it would be more efficient to
bundle this fix with other API updates, such as the need to rebase for incorporating an upstream fix related to the
Infoblox view parameter (https://github.com/kubernetes-sigs/external-dns/pull/3301) at a later point.

### Assume Role Strategy

Assuming a different AWS role could be implemented without ExternalDNS's built in `--aws-assume-role` as demonstrated in
https://github.com/openshift/external-dns-operator/pull/191 by using kiam, IRSA, or kube2iam. These are alternate

> methods for assuming a different AWS role, which include utilizing third-party software rather than interacting with

Reviewer: ```suggestion
methods for assuming a different AWS role, which require utilizing third-party software rather than interacting with
```

Author: Done in next push

ExternalDNS directly. Therefore, this method was not chosen.

### External ID

We considered supporting configuration of an [external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html)
for the ExternalDNS Operator when using an assumed role. ExternalDNS provides support for an external ID via the [`--aws-assume-role-external-id`](https://github.com/kubernetes-sigs/external-dns/blob/22da9f231dbc6faa3a668b507a4a06823a129609/pkg/apis/externaldns/types.go#L460)

> command line argument. AWS suggests the use of an external ID in specific scenarios to mitigate the confused deputy

Reviewer: Add a link for the `confused deputy` problem.  I thought you already did that?  Are you missing a commit?

Author: Yep, I caught it too while I was reviewing old comments. I thought I did it, but not sure what happened. Added.

as created by [AWS Shared VPC with Cross-account DNS Zones](/enhancements/installer/aws-cross-account-dns-zone.md) such
as [ipi-aws-pre-shared-vpc-phz](https://steps.ci.openshift.org/chain/ipi-aws-pre-shared-vpc-phz) and [ipi-aws-post-shared-vpc-phz](https://steps.ci.openshift.org/chain/ipi-aws-post-shared-vpc-phz).
These steps set up a shared VPC cluster with access to a cross-account DNS private hosted zone. However, we may choose to use
a public hosted zone instead, as there is no functional distinction between the two, and it simplifies the E2E testing
process.

### Graduation Criteria

This update is an extension of the existing v1beta1 ExternalDNS Operator API and does not impact the graduation
criteria for the v1beta1 API.

The ExternalDNS Operator updates to support Shared VPC are out of payload and are therefore not aligned with
OpenShift releases. Initially, the plan was to release this enhancement alongside of OCP 4.14. However, because
ExternalDNS Operator lacks STS support (see [Lack of STS Support](#lack-of-sts-support)), the decision was made to
complete implementation of this feature, but postpone the release until STS support is available.

> There is no requirement for backporting this feature, as outlined in [AWS Shared VPC with Cross-account DNS Zones](/enhancements/installer/aws-cross-account-dns-zone.md).

Reviewer 2: The phrasing here is ambiguous.  

* "'AWS Shared VPC with Cross-account DNS Zones' outlines that there **is no requirement** for backporting this feature."  (This is how I interpreted the statement initially.)
* "'AWS Shared VPC with Cross-account DNS Zones' outlines that there **is a requirement** for backporting this feature in OCP, but there **is no such requirement** for ExternalDNS Operator."  (This is what I think you must actually mean.)

Anyway, the reason why we are not backporting the feature in ExternalDNS Operator is that you can run a newer EDNSO version on an older OCP cluster, right?  It would be useful to state that explicitly.

Updates to the ExternalDNS Operator will be made available without being restricted by a feature gate. These updates
will be considered GA at the time of merging the feature.

#### Tech Preview -> GA

See [Dev Preview -> Tech Preview](#dev-preview---tech-preview).

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

* As outlined in [The Required Credentials Issue](#the-required-credentials-issue), when we bump the API from v1beta1,
  the conversion webhook will be required to handle the changes we make to the `Credentials` field. Nonetheless, in the
  initial implementation of this API, we will have no concern related to upgrading or downgrading.

### Version Skew Strategy

The updates to the ExternalDNS Operator are decoupled from the standard release schedule of OpenShift Container
Platform. Consequently, in a shared VPC cluster using cross-account DNS record creation with the Ingress Operator, there
is a possibility of installing a version of the ExternalDNS Operator that lacks support for this same feature. However,
it is important to note that the ExternalDNS Operator and the installer or Ingress Operator do not rely on each other's 

> logic. Hence, this mismatch does not pose a concern.

Reviewer 2: Usually this section discusses version skew between the component and API server, or possibly version skew between the operator and its operand.  I suppose there are no such concerns here, however, as you are adding a new API field where EDNSO is the consumer, and the new functionality in EDNSO is setting a flag that already exists in the external-dns images that we have been shipping.

### Operational Aspects of API Extensions

N/A

#### Failure Modes

- If the role ARN is not configured with the appropriate permissions, ExternalDNS will error and be unable to generate

>   DNS records.

Reviewer 2: It would be helpful to have an example `oc` command to get the logs and an example of the error message that could appear in those logs.

Author: Yup, Done.

#### Support Procedures

> TODO

Reviewer 2: Is this going to be done in a follow-up PR?

Author: Yea, I wasn't sure if this was important enough to do in the first round. I could do it as a follow-up PR, but I think if I don't do this now, it'll never get done.

I added small section.

## Implementation History

> Major milestones in the life cycle of a proposal should be tracked in `Implementation History`.

Reviewer 2: This looks like another to-do.

### Assume Role Strategy

Assuming a different AWS role could be implemented without ExternalDNS's built in `--aws-assume-role` as demonstrated in
https://github.com/openshift/external-dns-operator/pull/191 by using kiam, IRSA, or kube2iam. These are alternate
methods for assuming a different AWS role, which require utilizing third-party software in addition to ExternalDNS
(kiam, kube2iam) or the STS support (IRSA: bound service account token). Therefore, this method was not chosen.

### External ID

We considered supporting configuration of an [external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html)
for the ExternalDNS Operator when using an assumed role. ExternalDNS provides support for an external ID via the [`--aws-assume-role-external-id`](https://github.com/kubernetes-sigs/external-dns/blob/22da9f231dbc6faa3a668b507a4a06823a129609/pkg/apis/externaldns/types.go#L460)
command line argument. AWS suggests the use of an external ID in specific scenarios to mitigate privilege escalation
such as the [confused deputy problem](https://en.wikipedia.org/wiki/Confused_deputy_problem). However, for this effort,

> we opted not to include this functionality at the moment. TThe reason behind this decision is that it would become

Reviewer 2: ```suggestion
we opted not to include this functionality at the moment. The reason behind this decision is that it would become
```

necessary to incorporate support for external ID throughout all the components that enable shared VPC (including API,

> Install, and Ingress Operator), and there hasn't been a specific customer need for it thus far.

Reviewer 2: Why would it be necessary?  Do you mean that adding external id to the ExternalDNS API would require adding it to the other APIs for consistency?

In a future version of ExternalDNS Operator's API, we will resolve the requirement for the `Credentials` field through a
more appropriate solution as outlined in the alternative [Bumping the API](#bumping-the-api).

### Risks and Mitigations

N/A

### Drawbacks

#### Complexity

A minor drawback is the added complexity in supporting cross-account DNS hosted zones. This complexity involves handling
AWS resource sharing, managing IAM policies and roles, and understanding shared VPC architecture.

#### Security

In terms of security concerns, the IAM role is not considered sensitive because AWS incorporates security measures to
limit the usage of a role ARN to a particular AWS account.

#### Lack of STS Support

Another drawback for this feature is that the ExternalDNS Operator currently lacks support for the AWS Security Token
Service (STS). Given that STS is not supported, and the majority of ROSA (Red Hat OpenShift Service on AWS) customers
rely on STS, this shared VPC feature is not usable for most managed OpenShift users. The addition of STS support for
ExternalDNS Operator is currently under evaluation. However, until that is completed, the decision has been made to move

> forward with the implementation of shared VPC support.

Reviewer 3: These 2 chapters sound more as risks than drawbacks. WDYT?

### Bumping the API

The appropriate fix for addressing the problem outlined in [The Required Credentials Issue](#the-required-credentials-issue)
would be as follows:
1. Update the CRD schema to make the existing `Credentials` field optional.
2. Create a new version of the API that makes the optional field a pointer (assuming `Credentials` unset and `AssumeRole`
   set is a valid combination).
3. Update new API to not allow zero value of `Credentials`.
4. Make the conversion webhook convert new/nil to old/zero-value and old/zero-value to new/nil, and new/zero-value to
   old/zero-value.

However, the concern with bumping the API solely for this update is that the process of bumping can be quite
time-consuming, and it doesn't align the timeline of this enhancement. Additionally, it would be more efficient to
bundle this fix with other API updates, such as the need to rebase for incorporating an upstream fix related to the
Infoblox view parameter (https://github.com/kubernetes-sigs/external-dns/pull/3301) at a later point.

### Assume Role Strategy

Assuming a different AWS role could be implemented without ExternalDNS's built in `--aws-assume-role` as demonstrated in

> https://github.com/openshift/external-dns-operator/pull/191 by using kiam, IRSA, or kube2iam. These are alternate
> methods for assuming a different AWS role, which require utilizing third-party software rather than interacting with
> ExternalDNS directly. Therefore, this method was not chosen.

Reviewer 3: ```suggestion
https://github.com/openshift/external-dns-operator/pull/191 by using kiam, IRSA, or kube2iam. These are alternate
methods for assuming a different AWS role, which require utilizing third-party software in addition to ExternalDNS (kiam, kube2iam) or the STS support (IRSA: bound service account token). Therefore, these methods were not chosen.
```

Author: Done.
